platform/pulp/plugin/changeset/model.py

```diff
@@ -0,0 +1,182 @@
+        """
+        Args:
+            model (pulp.plugin.Content): A real content model object.
+
+        """
+        super(RemoteContent, self).__init__(model)
+        self._artifacts = set()
+        self.settled = False
+
+    @property
+    def key(self):
+        """
+        The natural key as a dictionary.
+
+        Returns:
+            dict: The natural key.
+        """
+        return {f.name: getattr(self.model, f.name) for f in self.model.natural_key_fields}
+
+    @property
+    def artifacts(self):
+        """
+        Associated artifacts (read only).
+
+        Returns:
+            frozenset: Associated artifacts.
+        """
+        return frozenset(self._artifacts)
+
+    def add_artifact(self, artifact):
+        """
+        Add a remote artifact.
+
+        Args:
+            artifact(RemoteArtifact): An artifact to add.
+        """
+        artifact.content = self
+        self._artifacts.add(artifact)
+
+    def update(self, model):
+        """
+        Update this `model` stored with the specified model that has been
+        fetched from the database. The artifacts are matched by `relative_path`
+        and their model object is replaced by the fetched model.
+
+        Args:
+            model (pulp.plugin.Content): A fetched content model object.
+        """
+        self.model = model
+        known = {a.model.relative_path: a for a in self._artifacts}
+        self._artifacts.clear()
+        for artifact in model.artifacts.all():
+            try:
+                found = known[artifact.relative_path]
+                found.model = artifact
+                self._artifacts.add(found)
+            except KeyError:
+                log.error(_('Artifact not matched.'))
+
+    def settle(self):
+        """
+        Settle matters that are prerequisite to associating the content
+        with a repository.  This mainly pertains to the artifacts being downloaded;
+        and associated artifact models saved; the content model saved.
+
+        Notes:
+            Called whenever an artifact has been downloaded.
+        """
+        for artifact in self._artifacts:
+            if not artifact.model.downloaded:
+                return
+        self.save()
+        self.settled = True
+
+    def save(self):
+        """
+        Save the content and related artifacts in a single DB transaction.
+        Due to race conditions, the content may already exist raising an IntegrityError.
+        When this happens, the model is fetched and replaced.
+        """
+        is_duplicate = False
+        with transaction.atomic():
+            try:
+                super(RemoteContent, self).save()
+            except IntegrityError:
+                is_duplicate = True
+            else:
+                for artifact in self._artifacts:
+                    artifact.model.content = self.model
+                    artifact.save()
+        if is_duplicate:
+            model = type(self.model)
```

Comment:
Now I'm especially getting confused about the use of "model" to refer to an instance or a class.